Tasks/Npm/npm.ts

```diff
@@ -0,0 +1,273 @@
+            tl.warning(tl.loc("NoBuildCredentials"));
+        }
+
+        return undefined;
+    }
+}
+
+async function main(): Promise<void> {
+    tl.setResourcePath(path.join(__dirname, 'task.json'));
+
+    let command = tl.getInput(NpmTaskInput.Command);
+    let workingDir = tl.getInput(NpmTaskInput.WorkingDir);
+
+    let findNpmrc = workingDir ? tl.findMatch(workingDir, ".npmrc") : [];
+    if (findNpmrc && findNpmrc.length > 1) {
+        throw Error(tl.loc("MultipleProjectConfigs"));
+    }
+    let projectNpmrc = findNpmrc[0];
+    pushNpmrc(projectNpmrc);
+
+    let npmrc = getTempNpmrcPath();
+    let npm = new NpmToolRunner(workingDir, npmrc);
+
+    let endpointId: string;
+    let feedId: string;
+    let registryLocation: string;
+    let npmRegistries: INpmRegistry[] = getLocalNpmRegistries(workingDir);
+    switch(command) {
+        case "install":
+            registryLocation = tl.getInput(NpmTaskInput.CustomRegistry);
+            endpointId = tl.getInput(NpmTaskInput.CustomEndpoint);
+            feedId = tl.getInput(NpmTaskInput.CustomFeed);
+            npm.line("install");
+            break;
+
+        case "publish":
+            registryLocation = tl.getInput(NpmTaskInput.PublishRegistry);
+            endpointId = tl.getInput(NpmTaskInput.PublishEndpoint);
+            feedId = tl.getInput(NpmTaskInput.PublishFeed);
+
+            // We delete their project .npmrc so it won't override our user .npmrc
+            tl.rmRF(`${workingDir}/.npmrc`);
+            npm.line("publish");
+            break;
+
+        case "custom":
+            registryLocation = tl.getInput(NpmTaskInput.CustomRegistry);
+            endpointId = tl.getInput(NpmTaskInput.CustomEndpoint);
+            feedId = tl.getInput(NpmTaskInput.CustomFeed);
+            npm.line(tl.getInput(NpmTaskInput.CustomCommand, true));
+            break;
+
+        default:
+            tl.setResult(tl.TaskResult.Failed, tl.loc("UnknownCommand", command));
+            return;
+    }
+
+    if (registryLocation == RegistryLocation.Feed) {
+        npmRegistries.push(NpmRegistry.FromFeedId(feedId));
+    } else if (endpointId) {
+        npmRegistries.push(NpmRegistry.FromServiceEndpoint(endpointId, command !== NpmCommand.Publish));
+    }
+
+    for (let registry of npmRegistries) {
+        if (registry.authOnly === false) {
+            appendToNpmrc(npmrc, `registry=${registry.url}\n`);
+        }
+
+        appendToNpmrc(npmrc, `${registry.auth}\n`);
+    }
+
+    await npm.exec();
+
+    tl.rmRF(npmrc);
+    popNpmrc(projectNpmrc);
+    tl.rmRF(getTempPath());
+}
+
+function getTempPath(): string {
+    let tempNpmrcDir
+        = tl.getVariable('Agent.BuildDirectory')
+        || tl.getVariable('Agent.ReleaseDirectory')
+        || process.cwd();
+    let tempPath = path.join(tempNpmrcDir, 'npm');
+    if (tl.exist(tempPath) === false) {
+        tl.mkdirP(tempPath);
+    }
+
+    return tempPath;
+}
+
+function getTempNpmrcPath(): string {
```

Comment:
I think call this getTempUserNpmrcPath, and the other methods saveProjectNpmrc and restoreProjectNpmrc or something like that